### PR TITLE
Cluster log replication

### PR DESCRIPTION
**Pull request by mikeb01**
Use archive replication to copy logs from leader to follower before beginning replay.

### CODE REVIEW

File: aeron-agent/src/main/java/io/aeron/agent/ClusterEventCode.java [hunk 1]
```diff
@@ -29,7 +29,9 @@
     ELECTION_STATE_CHANGE(1, ClusterEventDissector::dissectStateChange),
     NEW_LEADERSHIP_TERM(2, (eventCode, buffer, offset, builder) -> dissectNewLeadershipTerm(buffer, offset, builder)),
     STATE_CHANGE(3, ClusterEventDissector::dissectStateChange),
-    ROLE_CHANGE(4, ClusterEventDissector::dissectStateChange);
+    ROLE_CHANGE(4, ClusterEventDissector::dissectStateChange),
+    CANVASS_POSITION(5, ClusterEventDissector::dissectCanvassPosition),
+    REQUEST_VOTE(6, ClusterEventDissector::dissectRequestVote);
```
Comment (mjpt777) [inline aeron-agent/src/main/java/io/aeron/agent/ClusterEventCode.java:34]: Very useful.

File: aeron-agent/src/test/java/io/aeron/agent/ClusterLoggingAgentTest.java [hunk 2]
```diff
@@ -141,9 +142,11 @@ private void testClusterEventsLogging(
         Tests.await(WAIT_LIST::isEmpty);
 
         final Counter state = clusteredMediaDriver.consensusModule().context().electionStateCounter();
+
+        final Supplier<String> message = () -> ElectionState.get(state).toString();
         while (ElectionState.CLOSED != ElectionState.get(state))
         {
-            Tests.sleep(1);
+            Tests.wait(Tests.SLEEP_1_MS, message);
```
Comment (mjpt777) [inline aeron-agent/src/test/java/io/aeron/agent/ClusterLoggingAgentTest.java:149]: Should call is sleep and have message overload.

File: aeron-archive/src/main/resources/archive/aeron-archive-codecs.xml [hunk 3]
```diff
@@ -2,8 +2,8 @@
 <sbe:messageSchema xmlns:sbe="http://fixprotocol.io/2016/sbe"
                    package="io.aeron.archive.codecs"
                    id="101"
-                   version="5"
-                   semanticVersion="5.2"
+                   version="6"
+                   semanticVersion="5.3"
```
Comment (mjpt777) [inline aeron-archive/src/main/resources/archive/aeron-archive-codecs.xml:6]: `semanticVersion` is a FIX things and should not be changed.

File: aeron-cluster/src/main/java/io/aeron/cluster/Election.java [hunk 4]
```diff
@@ -666,6 +708,49 @@ private int followerReplay(final long nowNs)
...
+            }
+        }
+        else
+        {
+            workCount += logReplication.doWork();
+            if (logReplication.isDone())
+            {
+                // Which position should be updated here???
+                appendPosition = logReplication.replicatedLogPosition();
+                assert appendPosition == logReplicationPosition;
```
Comment (mjpt777) [inline aeron-cluster/src/main/java/io/aeron/cluster/Election.java:735]: We don't use `assert` in Java code. Either verify the position in business logic or not.

File: aeron-cluster/src/main/java/io/aeron/cluster/Election.java [hunk 5]
```diff
@@ -872,13 +961,16 @@ private void addCatchupLogDestination()
...
-        final String destination = ctx.isLogMdc() ? "aeron:udp?endpoint=" + thisMember.logEndpoint() : ctx.logChannel();
+        final String destination = ctx.isLogMdc() ?
+            "aeron:udp?endpoint=" + thisMember.logEndpoint() + "|alias=liveLog_" + thisMember.id() : ctx.logChannel();
         logSubscription.addDestination(destination);
         consensusModuleAgent.liveLogDestination(destination);
     }
 
     private Subscription addFollowerSubscription()
     {
+        assert logSessionId != CommonContext.NULL_SESSION_ID;
```
Comment (mjpt777) [inline aeron-cluster/src/main/java/io/aeron/cluster/Election.java:972]: Don't use assert.

File: aeron-cluster/src/main/java/io/aeron/cluster/ElectionState.java [hunk 6]
```diff
@@ -64,45 +64,49 @@
      */
     LEADER_READY(7),
 
+    LEADER_LOG_REPLICATION(8),
```
Comment (mjpt777) [inline aeron-cluster/src/main/java/io/aeron/cluster/ElectionState.java:67]: Needs javadoc.

File: aeron-cluster/src/main/java/io/aeron/cluster/ElectionState.java [hunk 7]
```diff
@@ -64,45 +64,49 @@
...
-    FOLLOWER_LOG_AWAIT(13),
+    FOLLOWER_LOG_AWAIT(14),
 
     /**
      * Publish append position to leader to signify ready for new term.
      */
-    FOLLOWER_READY(14),
+    FOLLOWER_READY(15),
+
+    FOLLOWER_LOG_REPLICATION(16),
```
Comment (mjpt777) [inline aeron-cluster/src/main/java/io/aeron/cluster/ElectionState.java:104]: Needs javadoc.

File: aeron-cluster/src/main/java/io/aeron/cluster/RecordingLog.java [hunk 8]
```diff
@@ -1385,6 +1391,29 @@ else if (-1 != snapshotIndex && -1 != logIndex)
...
+            {
+                throw new ClusterException("unknown recording id: " + replicatedRecordingId);
+            }
+
+            final long startPosition = -1 == snapshotIndex ?
+                recordingExtent.startPosition : snapshots.get(0).logPosition;
+
+            logRef.set(new Log(
+                replicatedRecordingId,
+                -1,
```
Comment (mjpt777) [inline aeron-cluster/src/main/java/io/aeron/cluster/RecordingLog.java:1407]: Use `NULL_VALUE` constant.

File: aeron-agent/src/test/java/io/aeron/agent/ClusterLoggingAgentTest.java (hunk 2 above)
Comment (mikeb01) [reply aeron-agent/src/test/java/io/aeron/agent/ClusterLoggingAgentTest.java]: Done, but I think we should look at removing the `Tests.sleep(long)` method.  We should always be providing context on what is being waited for and the current state on failure.  I'll pick this up as a background task to change existing usages.

**Pull request merged**